internal/metrics/binpack.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"context"
+	"math"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/binpack"
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+	"github.com/palantir/pkg/metrics"
+)
+
+type reportableEfficiency struct {
+	CPU    float64
+	Memory float64
+	GPU    float64
+}
+
+func reportableEfficiencyFromPackingEfficiency(efficiency *binpack.PackingEfficiency) reportableEfficiency {
+	return reportableEfficiency{
+		CPU:    efficiency.CPU,
+		Memory: efficiency.Memory,
+		GPU:    efficiency.GPU,
+	}
+}
+
+func reportableEfficiencyFromAvgPackingEfficiency(efficiency *binpack.AvgPackingEfficiency) reportableEfficiency {
+	return reportableEfficiency{
+		CPU:    efficiency.CPU,
+		Memory: efficiency.Memory,
+		GPU:    efficiency.GPU,
+	}
+}
+
+const (
+	packingEfficiencyMetricName = "foundry.spark.scheduler.packing_efficiency"
+
+	packingResourceTagKey = "foundry.spark.scheduler.packing_resource"
+	cpuTagValue           = "CPU"
+	memoryTagValue        = "Memory"
+	gpuTagValue           = "GPU"
+	maxTagValue           = "Max"
+
+	packingEfficiencyFunctionNameTagKey = "foundry.spark.scheduler.packing_function"
+
+	// non-existing node to explicitly report average packing efficiency
+	avgEfficiencyNodeName = "average"
+)
+
+var (
+	cpuTag    = metrics.MustNewTag(packingResourceTagKey, cpuTagValue)
+	memoryTag = metrics.MustNewTag(packingResourceTagKey, memoryTagValue)
+	gpuTag    = metrics.MustNewTag(packingResourceTagKey, gpuTagValue)
+	// represents higher of CPU and Memory packing efficiencies. GPU is explicitly excluded for now
+	maxTag = metrics.MustNewTag(packingResourceTagKey, maxTagValue)
+)
+
+// ReportPackingEfficiency report packing efficiency metrics for a single packing result.
+func ReportPackingEfficiency(
+	ctx context.Context,
+	packingFunctionName string,
+	nodesSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	packingResult *binpack.PackingResult) {
+
+	contextTags := metrics.TagsFromContext(ctx)
```

Comment:
I think its unlikely that we'll want all of these tags, the cardinality of the metric will be massive

Reply:
Do I understand correctly that the important tags, such as instance-group, deployment are populated automatically when calling `metrics.FromContext(ctx)...`?